EuiTableOfRecords component

uboness opened the pull request:
`EuiTableOfRecords` is a high level component that aims to easy the way one creates a table of... (wait for it....) records!!!

What is a record? Think of a record in a data store - typically it represents an entity with a very clear schema. It is something that can be presented in a tabular form.

The goal of this high level component is to make the consumer not think about design. Instead, the consumer only need to define the functional requirements of the table (features), the data, and the type of interaction the user should have with the shown records.

This high level component is as stateless as it needs to be. Meaning, all the management of the data (where it's coming from), record selection and loading pages... all of these are expected to be managed externally to this component. Typically one would use a container component to wrap around this table that will either manage this state internally to that component, or use other state stores such as Redux.

This initial commit includes:

- A first implementation of the `EuiTableOfRecords`
- Support for pagination
- Support for record selection
- Support for custom rendering of the record data (introducing "value renderers")
- Supports adding actions per record

This commit also comes with a dedicated documentation section in the EUI doc site.

cjcenizal commented on src/components/value_renderer/health/health.js:
```diff
@@ -0,0 +1,17 @@
+    const color = config.color(value, ctx);
+    return <EuiHealth color={color}>{resolvedContent}</EuiHealth>;
+  };
+};
```
What's the advantage of using this renderer over the `EuiHealth` component directly?

```jsx
          render: (value) => {
            const color = value === 0 ? 'danger' : 'success';
            return (
              <EuiHealth color={color}>{value}</EuiHealth>
            );
          },
```

uboness replied:
not a huge advantage... but in the same line as in `link()` or using descriptors for buttons... having this as the default goto will promote consistency (e.g. consumers won't immediate start thinking about providing their own components for representing health status)... maybe a better name would be "status" rather than health... as it's more the idea we want to convey rather than talk about the actual implementation... yea... the more I think about it (write these lines) the more I like "status" :)

cjcenizal commented on src/components/value_renderer/link/link.js:
```diff
@@ -0,0 +1,17 @@
+    const onClick = () => config.onClick(value, ctx);
+    return <EuiLink type={config.type} color={config.color} onClick={onClick}>{resolvedContent}</EuiLink>;
+  };
+};
```
Same here, why not just use `EuiLink` directly?

cjcenizal commented on src/services/sort/index.js:
```diff
@@ -1 +1,4 @@
 export { SortableProperties } from './sortable_properties';
+export { SortDirectionType, SortDirection } from './sort_direction';
+export { PropertySortType } from './property_sort';
+export { Comparators } from './comparators';
```
These new services seem to duplicate the functionality of `SortableProperties`. Is there any way we can use the existing service?

uboness replied:
tbh I'm really not a big fan of SortableProperties (it ties to a discussion we recently had about the meaning of the extra arrows). We can simplify things quite a bit here... we only sort on one field at a time, so no need to keep state for all the other sorts... the "sortable" indicator needs to be a different icon than what we have today (and then you don't have the need for keeping the extra state).

Look how simple`SortDirection` is when you don't need state compared to `SortableProperties`...

cjcenizal commented:
I forgot to also mention I found a few other issues:

## Row action accessibility

I wasn't able to tab to the row actions, which means they're inaccessible to people who use the keyboard to navigate. We can solve this by leaving the actions in the DOM and using CSS to set their opacity to 1 when the row is hovered or the button is focused.

## Search and bulk actions

Were you planning on also bundling the search input and bulk actions into this component? I think these are worth adding.

## Table states

We'll need to support loading, error, and empty states with our tables. We'll need input from @cchaos and @snide on how these should appear. I think a good, safe choice would be to simply allow the consumer to pass in a React component to a `holdingState` prop, which can be specified at any time for any reason. This would be a good catch-all for all loading, error, and empty states, and any additional ones which arise in the future.

## Sorting behavior

Can we standardize our sorting behavior based on https://github.com/elastic/kibana/issues/9865? This is something provided by the `SortableProperties` service.

uboness commented on src-docs/src/views/table_of_records/multiple_record_actions.js:
```diff
@@ -0,0 +1,206 @@
+
+  constructor(props) {
+    super(props);
+    this.state = this.computeState({
```
it matters not for this component, so we're not trying to mimic anything. As far as the consumer is concerned this component is stateless, meaning you need to provide all the necessary info for it to work as props. For this reason, the data can be loaded from the server async, or it can be stored in memory... but for this component it really doesn't matter therefore focusing on this question makes little sense to me

uboness commented on src-docs/src/views/table_of_records/multiple_record_actions.js:
```diff
@@ -0,0 +1,206 @@
+  }
+
+  computeState(criteria) {
+    const page = loadPage(criteria.page.index, criteria.page.size, criteria.sort);
```
sure, we can... I don't mind... but again, this is the least important part of the example one should focus on

uboness commented on src-docs/src/views/table_of_records/multiple_record_actions.js:
```diff
@@ -0,0 +1,206 @@
+        },
+        sort: criteria.sort
+      }
+    };
```
oh... but do see a reason :)... it has to do with the fact that with this example, we're using the component's state to store the necessary state for the table... loading the data however is a completely separated concerns. I think it's important to separate these two concerns and not promote tying them to one another.

uboness commented on src-docs/src/views/table_of_records/multiple_record_actions.js:
```diff
@@ -0,0 +1,206 @@
+  deletePerson(personToDelete) {
+    const i = people.findIndex((person) => person.id === personToDelete.id);
+    if (i >= 0) {
+      people.splice(i, 1);
```
I think we need to discuss the nature of this component as I gather from the last few questions (including this one) that it's totally not clear. I tried to explain it in the "What is a high level component" call out section.. perhaps I did not do a good enough job.

In a nutshell.. this high level component doesn't (nor should) care about where the data is managed, whether we hold it in memory or wether we load it remotely... or wether there's a delay in loading it. It is the least important part of the examples here. The important part is the props you pass to the component and the callback you register. for the rest, there can be many ways to deal with data... but this is really not the point here.

In this specific example, we want to show how to register an action... this action happens to be a "delete person" action... I could have shown an action that just shows an alert dialog with a "hi there ${person.name}" message.... but I didn't... I wanted to make it more realistic so I implemented a "deletePerson" action... now because in the example we hold all people in an array in memory, the implementation is a simple splice... but really... who cares... it could have been something completely different...

uboness commented on src-docs/src/views/table_of_records/multiple_record_actions.js:
```diff
@@ -0,0 +1,206 @@
+
+    };
+
+    return <EuiTableOfRecords config={config} model={this.state}/>;
```
separating the callback doesn't work well... I had it separated initially and moved away from that. The main reason is that pagination, sorting and (later) filtering are all facets of the same concerns, which is basically defining the criteria of the data you're viewing. How does the table behave when you sort? do you clear selection - do you reset pagination? how does the table behave when you pagination -  do you clear selection do you keep the sorting? these are all questions you want to avoid consumer asking or even dealing with. This behaviour needs to be determined and dictated by the component itself - this is the only way we can create consistent behaviour across the board - it's not just about documenting our design and UX patterns, it's also about trying to enforce them whenever we can... and if we can't enforce them, then at least promote them. Moving away from these patterns should be an exception that requires addition work and should be well documented why its done in the first place. To achieve this, you need to have a single callback that notifies about a change in the "criteria"... consumers don't know what drove this change, nor they should care about it... all they know is that the table is "requesting" to update the data based on a new criteria (it could be that the user sorted and because of that the table also decided to reset pagination... or that the user added a filter which caused a pagination reset as well... but you don't know... and it's ok... these are the rules and behaviours the table defines and takes care of for you).

personally I don't think `isFetching` should be the concern of the table... it doesn't really know what "fetching is"... if you use an in-memory model around it, it's meaningless... if you use async/remote model around it it has a meaning... is fetching is a higher level concern... higher than this table.

The flattening of this model breaks it's high level and brings it lower than intended. At a high level you deal with "data" which is: 1. a list of records, and 2. the criteria that describes these records. This is what the table cares about - data... not rows. Again... I feel we need to have a discussion about the goal of the component in the first place and how the users/consumers should look at it.... I feel there's a fundamental miss in the overall intention here, that (I guess) I miss out when trying to explain it.

uboness replied:
separating the callback doesn't work well... I had it separated initially and moved away from that. The main reason is that pagination, sorting and (later) filtering are all facets of the same concerns, which is basically defining the criteria of the data you're viewing. How does the table behave when you sort? do you clear selection - do you reset pagination? how does the table behave when you pagination -  do you clear selection do you keep the sorting? these are all questions you want to avoid consumer asking or even dealing with. This behaviour needs to be determined and dictated by the component itself - this is the only way we can create consistent behaviour across the board - it's not just about documenting our design and UX patterns, it's also about trying to enforce them whenever we can... and if we can't enforce them, then at least promote them. Moving away from these patterns should be an exception that requires addition work and should be well documented why its done in the first place. To achieve this, you need to have a single callback that notifies about a change in the "criteria"... consumers don't know what drove this change, nor they should care about it... all they know is that the table is "requesting" to update the data based on a new criteria (it could be that the user sorted and because of that the table also decided to reset pagination... or that the user added a filter which caused a pagination reset as well... but you don't know... and it's ok... these are the rules and behaviours the table defines and takes care of for you).

personally I don't think `isFetching` should be the concern of the table... it doesn't really know what "fetching is"... if you use an in-memory model around it, it's meaningless... if you use async/remote model around it it has a meaning... is fetching is a higher level concern... higher than this table.

The flattening of this model breaks it's high level and brings it lower than intended. At a high level you deal with "data" which is: 1. a list of records, and 2. the criteria that describes these records. This is what the table cares about - data... not rows. Again... I feel we need to have a discussion about the goal of the component in the first place and how the users/consumers should look at it.... I feel there's a fundamental miss in the overall intention here, that (I guess) I miss out when trying to explain it.

uboness commented on src-docs/src/views/table_of_records/table_of_records_example.js:
```diff
@@ -0,0 +1,423 @@
+        <h2>TableOfRecords</h2>
+      </EuiTitle>
+      <p>
+        <EuiCode>EuiTableOfRecords</EuiCode> is a high level component that aims to simplify and unifiy the way
```
another example in the gap of thinking :)... lets talk about it..

uboness commented:
> I also think we should try to slim down the documentation. It's very thorough but too much documentation risks overwhelming the reader and thus not being read. I made some comments about ways we can do this.

Personally I don't agree... I think that the docs need to convey how one can use a component and not assume everything from one example (can I do X... can I do Y... one example assumes the user knows all the options and the required props you must provide or the optional/required feature a component supports). I personally like having a build up where you start with the bare minimum and you enhance things gradually. You still provide ppl with an option to directly jump to the most "featured" example... but you also let them see they can use it in different ways. 

This doesn't really tie to this component though, this is something we need to decide on a much higher level for this framework as a whole. I'd love to hear @snide 's opinion on it, but also other kibana engs that eventually use it and are the target audience for this docs.

uboness commented:
@cjcenizal thanks for taking the time to start the review... exciting times... I'm in the process of answering a lot of your questions... it'll take some time though :)

uboness commented:
> Row action accessibility

Absolutely... Known issue... Discussed with Dave about it... I know how to fix this 🙂

> search and bulk actions 

Absolutely... In a follow-up pr once this is in (it gotten quite big already)

> table states 

Absolutely... In a follow-up pr

cjcenizal commented:
To support the ability to use `dataType` as well as append a `data-test-subj` attribute to a cell, maybe we can support a `props()` callback alongside the `render()` callback. You can only use one or the other. In the `props()` callback you can return a `props` object which will be passed to the `td` element.

uboness commented on src/components/value_renderer/default_renderer/defaultRenderer.js:
```diff
@@ -0,0 +1,29 @@
+import { text } from '../text';
+import { booleanText } from '../boolean';
+
+export const defaultRenderer = (value) => {
```
This is used internally by the components as the default renderer when no rendering hints or customisation is provided by the consumer. without it, every time you'd want to have defaults you'll find yourself if-elsing repetitively - this a one stop shop for default value rendering.

cjcenizal commented:
I've submitted a PR with some suggestions: https://github.com/uboness/eui/pull/1/

cjcenizal commented:
@uboness I noticed a bug with sorting... not sure what's causing this, but it looks like the direction of the arrow is the reverse of the sort direction (ascending === A-to-Z === arrow points up):

![image](https://user-images.githubusercontent.com/1238659/35177561-6871c81e-fd35-11e7-8152-069379d53533.png)

uboness commented:
Hmm... It's actually intentional... I mean, I didn't look at common/other  implementations... It just made sense to me that asc is down... Even looking at the screenshot above makes total sense to me. But if we need to change that, we can of course

uboness commented on src/components/table_of_records/table_of_records.js:
```diff
@@ -0,0 +1,752 @@
+
+  togglePopover(id) {
+    this.setState((prevState) => ({
+      ...prevState,
```
strange... I was sure that the merge only happens when you provide a value to `setState`, and that it's your responsibility to merge when providing a callback.. but it looks like I was wrong:

> Both prevState and props received by the updater function are guaranteed to be up-to-date. The output of the updater is shallowly merged with prevState.

will change

uboness commented on package.json:
```diff
@@ -23,12 +23,14 @@
     "brace": "^0.10.0",
     "classnames": "^2.2.5",
     "core-js": "^2.5.1",
+    "date-fns": "^1.29.0",
```
I thought about that... wasn't sure if we plan to continue with moment or not (I remember some debate around it... if the plan is to continue using it... sure, I'll update, but if the plan is to slowly remove `moment.js`, then we better not... hmm... 

@kjbekkelund / @epixa what's the status of moment.js in Kibana? is it safe to use?

uboness replied:
got the green light to use moment.js... will make the change

uboness replied:
@chrisronline done

chrisronline replied:
Sweet, thanks!

uboness commented on src/components/table_of_records/table_of_records.js:
```diff
@@ -0,0 +1,752 @@
+      const checked = this.state.selection && this.state.selection.length > 0;
+      const onChange = (event) => {
+        if (event.target.checked) {
+          const selectableRecords = model.data.records.reduce((records, record) => {
```
heh... no sure why I didn't do that in the first place :) 👍

uboness commented on src/components/table_of_records/table_of_records.js:
```diff
@@ -0,0 +1,752 @@
+import React from 'react';
+import * as _ from 'lodash';
```
👍

uboness commented on src/components/table_of_records/table_of_records.js:
```diff
@@ -0,0 +1,752 @@
+import React from 'react';
```
I thought about it a few times... but always came back to the same conclusion. The thing that makes this file large is the rendering logic and it feels awkward to break it apart. Personally I like having all rendering logic in one file... but if ppl insist I can break it apart a bit (not too much though... maybe to `table`, `footer` and later `header`).... but again.. personally I'd vote for keeping all rendering logic in one place

bmcconaghy replied:
Yeah I was thinking table header and footer as the breakout.

chrisronline commented on src/components/table_of_records/table_of_records.js:
```diff
@@ -0,0 +1,752 @@
+        size
+      }
+    };
+    this.props.config.onDataCriteriaChange(criteria);
```
I'm going to think about it more, but this feels weird to me. I'm fairly sure I understand the _what_ and _why_, but it feels awkward to remember to tell someone else that the criteria changed, but when it _seems_ like the only thing that will care is the `TableOfRecords` itself? Is this just so the  consuming component can manage the list? Am I making any sense? :P

uboness replied:
:)... I can totally understand how you feel about it, and you're probably missing context here... so here it goes... 

The point with this table was to make it as stateless as possible. The main reasons for this is that it should be possible to use it in any environment regardless of where and how one stores state. 

Some state, is totally internal to the table (selection is a good example) and with this type of state we indeed keep it within the table itself and don't leak it out (the user can still register a callback method to "listen" to selection changes, but it's optional).

With data, things are a bit different. Data is not typically help in memory and one can manage it it many ways. For example, one can have a higher level component (container component) that stores data (or some of it) in it's own internal state. Alternatively one can use a redux store to store the state. This is where this specific table stops making the decisions for the user. I want to enable ppl to use this table with redux or however they choose to store the state.

That said, we do plan to provide a higher level container on top of this table that will handle *most* of the state for the user. It could be very useful to have this component, specially for K6 (reduxless env) and also for use cases where one actually wants to hold all data in memory. The plan was to add that container (currently named `TableOfRecordsDS`) in a later commit, once this goes in.

> but when it seems like the only thing that will care is the TableOfRecords itself?

you'll get that with `TableOfRecordsDS`... well... mostly... my plan is that it'll be flexible enough to either hold everything in memory (in its state), or delegate the data loading to an async function returning a promise... on of the two

>  Is this just so the consuming component can manage the list?

either the consuming component or perhaps a redux store... or something else... trying not to make it the concern of this table.

> Am I making any sense? :P

yes, you do :).... it all started when I came from the world of redux and best practices there (one things we found out that with redux it's best to work with stateless components and let the state be stored in a single place rather than scatter it all over the place). So it was important for me to make sure this works well in that env.... but I totally recognize the need for a component that will be "fully managed" or "mostly managed"... hence `TableOfRecordsDS`

btw... I need help with finding a good name for `TableOfRecordsDS` 🤦‍♂️

chrisronline replied:
Gotcha, yea that totally makes sense. I totally agree about adding another HoC to facilitate a couple common use cases of managing the records. Thanks for clarifying!

uboness commented:
@cjcenizal I'm back on the fence wrt requiring the `pagination` config to enable pagination. The main reason for this:

1. There should only be one thing that enables a feature.. the minute you start sending ppl to change multiple things it because a bit of a mess
2. code-wise we now need to validate proper use by making sure that if the model indicates paging but the table isn't configured with pagination we should throw an error - it's a misuse and/or confusion with the consumer.
3. pagination needs to be enabled by default... this is not possible now that we require `config.pagination` to be configured. We could solve that by setting assuming pagination and letting the user to opt-out of it using `config.pagination.enabled: false` setting... but that pretty much brings us to the beginning - I much rather tell the user "you want to opt out of pagination? don't provide the page criteria" that's it.

`model.criteria.page` can be the only thing that enables/disabled pagination.... for further configuration of it (that is... configuring how the pagination behaves) we can expose as `config.pagination` setting (a la page size options)

cjcenizal commented:
@uboness Hehe this is why I wanted to break up and recombine the parts of configuration and model, so we could group things into cohesive objects, e.g. a single `pagination` object which contains all relevant props. Then we could just add/remove that to enable/disable pagination.

Why does pagination need to be enabled by default, btw?

uboness commented:
> @uboness Hehe this is why I wanted to break up and recombine the parts of configuration and model, so we could group things into cohesive objects, e.g. a single pagination object which contains all relevant props. Then we could just add/remove that to enable/disable pagination. Aside from that, the benefit of consolidating the criteria facts under one object is consistency - we expose the `onModelCriteriaChange` event where the consumer get the criteria object... it's somewhat simple to think as this object as an atom where the same criteria object is also provided back to the table. Note, I didn't have `onModelCriteriaChange` and instead I indeed broke things by feature (e.g the user could listen in to pagination events with `onPaginationChange` callback within the `pagination` object... but that was a problematic approach - it doesn't provide us full control over the behaviour of the table (for example, we want to make sure pagination is reset when the user sorts... we can only do that if we have expose both sorting & pagination in a single object with a single callback... the alternative will mean there'll be multiple requests and flickery UI)

also... it'd feel strange that the page criteria will be apart from the data... these two really belongs together IMO... you have the data and what describes the data in one place.

> Why does pagination need to be enabled by default, btw?

I stated based on the places I already foresee this component being used you'd want pagination in all of them... so it feels more like an opt-out feature rather then an opt-in one. Even for those use cases that I do not foresee (yet)... as long as we don't support infinite scrolling, or something like that, I think for proper UX you'd **almost** always want pagination. But regardless of that, if we only rely on the page criteria to enable disable pagination, it's no longer a matter of whether the user opts in/out of it... if the user attaches page criteria we'll enable it... if not, we won't... simple

uboness commented on src/components/value_renderer/compound/join.js:
```diff
@@ -0,0 +1,5 @@
+import { isArray } from 'lodash';
+
+export const join = (renderer, delim = ', ') => {
+  return (value) => isArray(value) ? value.map(item => renderer(item)).join(delim) : renderer(value);
```
lodash's join is not aware of our rendering defaults... the advantage of having this as a renderer is that it is a composite of other renderers. This enables use to create a hierarchy of renderers all following our defaults (see [defaultRenderer](https://github.com/uboness/eui/blob/d1bd6cb2f5c1f0bd8ea898000cc835e762f498ec/src/components/value_renderer/default_renderer/defaultRenderer.js#L25))

uboness commented on src/components/value_renderer/compound/property.js:
```diff
@@ -0,0 +1,5 @@
+import { get } from 'lodash';
+
+export const property = (prop, renderer) => {
+  return (value) => renderer(get(value, prop));
```
same as above... "composition"

uboness commented:
> @uboness I noticed a bug with sorting... not sure what's causing this, but it looks like the direction of the arrow is the reverse of the sort direction (ascending === A-to-Z === arrow points up):

fixed... the issue was in the comparators (added tests).... I wonder if we could use a better icon to indicate sort direction... something like:

![sort](https://cdn4.iconfinder.com/data/icons/ui-controls-editor/32/interface-controls-sort-ascending-256.png)

chrisronline commented:
@cjcenizal Feels lateral to me. I don't honestly have a strong preference either way.

uboness commented:
@cjcenizal I've pushed the changes we worked on yesterday. Added tests for all the new components I created and also updated the docs. As we said, you still need to look at the `updateFocus()` method we disabled in context menu.

@bmcconaghy cj wills till work on fixing the build (we had to comment out the `updateFocus` method in context menu as it was in the way for properly handling focus/blur events). That said, I split the code as discussed and a bit more than just the footer (which is now named `PaginationBar`... also split the code for the record actions. overall I think it's a good balance (as mentioned somewhere above, I'm not keen on splitting the rendering of table itself)

kjbekkelund commented on src-docs/src/views/table_of_records/implicit_record_action.js:
```diff
@@ -0,0 +1,165 @@
+          description: `Person's nickname / online handle`,
+          render: EuiValueRenderers.link({
+            onClick: (value) => {
+              window.open(`http://www.github.com/${value}`, '_blank');
```
This breaks accessibility, I think: https://developer.mozilla.org/en-US/docs/Web/API/Window/open#Never_use_<a_href_onclickwindow.open(...)>

I know this is our docs and not Kibana itself, but we likely shouldn't use bad patterns in our docs either.

uboness replied:
will be fixed soon
